src/nodes/hypertable_modify.c

```diff
@@ -747,18 +747,22 @@ ExecModifyTable(CustomScanState *cs_node, PlanState *pstate)
 	 */
 	if ((operation == CMD_DELETE || operation == CMD_UPDATE) && !ht_state->comp_chunks_processed)
 	{
-		if (ts_cm_functions->decompress_target_segments)
+		/* Modify snapshot only if something got decompressed */
+		if (ts_cm_functions->decompress_target_segments &&
+			ts_cm_functions->decompress_target_segments(ht_state))
```

Comment:
OK, I think I managed to understand the rationale why we add the `RegisterSnapshot` below. But what about this new check, why is it needed? Your test passes without this change with just the `RegisterSnapshot`. I also double-checked that it fails w/o the `RegisterSnapshot`.

Reply:
It seems reasonable to me that the update process should work even if we just increment command counter and take the new snapshot unconditionally always.

Reply:
Well, that seems to be the correct way to work with the snapshots there, so maybe we should instead apply it unconditionally? If you only apply it under some complicated conditions, the bugs will be more difficult to find. E.g. now it looks like we should add isolation tests with both REPEATABLE READ and READ COMMITTED to test the `result == TM_Deleted && !IsolationUsesXactSnapshot()` code path. We still have to do the potentially problematic snapshot operations in some cases, and if we add more conditions to invoke them, I would say it doesn't really improve the things.